extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/CoroutineInvocationHandler.kt

```diff
@@ -1,25 +1,43 @@
 package org.jboss.resteasy.reactive.server.runtime.kotlin
 
-import kotlinx.coroutines.GlobalScope
+import io.vertx.core.Vertx
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.launch
+import org.eclipse.microprofile.context.ManagedExecutor
 import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext
 import org.jboss.resteasy.reactive.server.spi.EndpointInvoker
 import org.jboss.resteasy.reactive.server.spi.ServerRestHandler
+import org.slf4j.LoggerFactory
+
+private val logger = LoggerFactory.getLogger(CoroutineInvocationHandler::class.java)
+
+class CoroutineInvocationHandler(private val invoker: EndpointInvoker,
+                                 private val coroutineScope: CoroutineScope,
+                                 managedExecutor: ManagedExecutor) : ServerRestHandler {
+
+    private val executorDispatcher = ExecutorDispatcher(managedExecutor)
 
-class CoroutineInvocationHandler(private val invoker: EndpointInvoker) : ServerRestHandler {
     override fun handle(requestContext: ResteasyReactiveRequestContext) {
         if (requestContext.result != null) {
             return
         }
+        if (invoker !is CoroutineEndpointInvoker) {
+            requestContext.handleException(IllegalStateException("Not a coroutine invoker"), true)
+            return
+        }
+
+        val dispatcher: CoroutineDispatcher = Vertx.currentContext()?.let(::VertxDispatcher)
```

Comment:
I don't understand how this selects between `@Blocking` and not?

Reply:
This one was tested empirically. When I defined `@Blocking` endpoints, the vertx context was null and it was running one of the managed executor threads.
Maybe there is a better way to make the distinction?

Reply:
Perhaps we should just disallow the case of `@Blocking` when a suspending function is being used as this use case would not yield any runtime benefit I can think of

Reply:
On one side I agree with you @geoand. You don't use reactive when you have blocking code.

On the other side, most applications have some kind of blocking function call. In that case, it needs to be made on thread that can block. It would be very nice to have built-in support for that instead of leaving it to the user.

Reply:
My point is more that if we do allow this, users will think they are gaining some kind of runtime benefit, when in fact that are gaining none and instead the lines of blocking vs non-blocking just become more blurry.

Reply:
That already happens with the current resteasy reactive extension right?

Reply:
Does what already happen?

Reply:
The fact that you can add the `@Blocking` annotation to a normal endpoint. Resteasy reactive extension already supports it and runs the code on a different thread.
I was planning to do the same with suspending endpoints to be consistent.

Reply:
Yes, that is supported by IMHO, it doesn't make sense for suspended functions. @FroMage WDYT?

Reply:
IMHO this should always just dispatch back to the vert.x executor. One potential issue is that you can have situations where something earlier in the chain has dispatched to a worker (e.g. Security JPA), so you are not currently on the IO thread even though this is where we want to execute. In this situation we would need to dispatch back I think. 

I think for now we should just be running everything on the IO thread, and if there is demand for @Blocking coroutines we can revisit it, but I don't really think they make sense at this point.

Reply:
Ok, so we can state we don't support `@Blocking`, i.e. will have unpredictable behavior.
In this case, can I assume the request needs to come from Vertx?
```suggestion
        val dispatcher: CoroutineDispatcher = Vertx.currentContext()?.let(::VertxDispatcher) ?: throw IllegalStateExcetion()
```
Because currently, this can happen. I'm just not sure if it only happens with `@Blocking` or some other feature.

Reply:
We can add simply prevent the use of `@Blocking` with corroutines at build time so you won't need to worry about it at runtime.
This can be done in a follow up

Reply:
Agreed.
I've blocked usages of `@Blocking` (pun intended) annotations with suspending functions.
We can add support later.

Reply:
Agreed, I don't see how it can make sense to want a coroutine on a fresh thread (even if it's really a thread from a thread pool): they're meant to be reusable for fast suspend, which imply IO threads. Let's detect and disallow at build time.

Reply:
I'll do a follow-up once this is in unless @pak3nuh has something ready to go and wants to update the PR